File: lessons/compiled_lang/rust/src/o10.options.rs
use std::env;
use std::process::exit;

fn usage(script_name: &str) -> String {
    format!("\nUsage: {script_name} [-c] [-e] [-l] [-k] [-p] [-m] [-t] [-h|-?]\n\n  -c  Coffee\n  -e  Espresso\n  -l  Latte\n  -k  Machiato\n  -p  Capucino\n  -m  Mocha\n  -t  Tea\n  -h  Display this help message\n  -?  Display this help message\n\n")
}

fn main() {
    let all_args: Vec<String> = env::args().collect();
    let script_name = &all_args[0];
    let args = &all_args[1..];

    // Manual args() iteration naturally preserves the command-line order,
    //  unlike a hash/map-backed collection - needed since the expected
    //  output lists orders in exactly the order given.
    let mut orders: Vec<&str> = Vec::new();
    for arg in args {
        match arg.as_str() {
            "-c" => orders.push("coffee"),
            "-e" => orders.push("espresso"),
            "-l" => orders.push("latte"),
            "-k" => orders.push("macchiato"),
            "-p" => orders.push("capucino"),
            "-m" => orders.push("mocha"),
            "-t" => orders.push("tea"),
            "-h" | "-?" => { print!("{}", usage(script_name)); exit(0); }
            _ => {}
        }
    }

    if orders.is_empty() {
        eprint!("{}", usage(script_name));
        exit(1);
    }

    println!();
    println!("You ordered: ");
    for drink in orders {
        println!("* {drink}");
    }
}

File: lessons/compiled_lang/go/src/o10.options.go
package main

import (
	"fmt"
	"os"
)

func usage(out *os.File, scriptName string) {
	fmt.Fprintln(out)
	fmt.Fprintln(out, "Usage: "+scriptName+" [-c] [-e] [-l] [-k] [-p] [-m] [-t] [-h|-?]")
	fmt.Fprintln(out)
	fmt.Fprintln(out, "  -c  Coffee")
	fmt.Fprintln(out, "  -e  Espresso")
	fmt.Fprintln(out, "  -l  Latte")
	fmt.Fprintln(out, "  -k  Machiato")
	fmt.Fprintln(out, "  -p  Capucino")
	fmt.Fprintln(out, "  -m  Mocha")
	fmt.Fprintln(out, "  -t  Tea")
	fmt.Fprintln(out, "  -h  Display this help message")
	fmt.Fprintln(out, "  -?  Display this help message")
	fmt.Fprintln(out)
}

func main() {
	scriptName := os.Args[0]
	args := os.Args[1:]

	if len(args) == 0 {
		usage(os.Stderr, scriptName)
		os.Exit(1)
	}

	var orders []string
	for _, arg := range args {
		switch arg {
		case "-c":
			orders = append(orders, "coffee")
		case "-e":
			orders = append(orders, "espresso")
		case "-l":
			orders = append(orders, "latte")
		case "-k":
			orders = append(orders, "macchiato")
		case "-p":
			orders = append(orders, "capucino")
		case "-m":
			orders = append(orders, "mocha")
		case "-t":
			orders = append(orders, "tea")
		case "-h", "-?":
			usage(os.Stdout, scriptName)
			os.Exit(0)
		default:
			usage(os.Stderr, scriptName)
			os.Exit(1)
		}
	}

	fmt.Println()
	fmt.Println("You ordered: ")
	for _, drink := range orders {
		fmt.Println("* " + drink)
	}
}
